From 27af9cfadbd80fdb011a94b59496e61830ff1988 Mon Sep 17 00:00:00 2001
From: Anders Englund <y94anden@hotmail.com>
Date: Sat, 17 Aug 2019 22:43:54 +0200
Subject: [PATCH] Added support for RS485

Shamelessly stolen from https://github.com/SodaqMoja/optiboot

Seems to build smaller than 512 bytes if not including EEPROM.
---
 optiboot/bootloaders/optiboot/Makefile   |   7 +
 optiboot/bootloaders/optiboot/optiboot.c |  51 ++-
 optiboot/bootloaders/optiboot/pin_defs.h | 547 +++++++++++++++++++++++++++++++
 3 files changed, 596 insertions(+), 9 deletions(-)

diff --git a/optiboot/bootloaders/optiboot/Makefile b/optiboot/bootloaders/optiboot/Makefile
index 3f923a8..b844425 100644
--- a/optiboot/bootloaders/optiboot/Makefile
+++ b/optiboot/bootloaders/optiboot/Makefile
@@ -229,6 +229,12 @@ dummy = FORCE
 endif
 endif
 
+HELPTEXT += "Option RS485=B1              - use specified PIN for RS485 dir\n"
+ifdef RS485
+RS485_CMD = -DRS485=$(RS485)
+dummy = FORCE
+endif
+
 HELPTEXT += "Option SINGLESPEED=1         - do not use U2X mode on UART\n"
 ifdef SINGLESPEED
 ifneq ($(SINGLESPEED), 0)
@@ -248,6 +254,7 @@ COMMON_OPTIONS = $(BAUD_RATE_CMD) $(LED_START_FLASHES_CMD) $(BIGBOOT_CMD)
 COMMON_OPTIONS += $(SOFT_UART_CMD) $(LED_DATA_FLASH_CMD) $(LED_CMD) $(SS_CMD)
 COMMON_OPTIONS += $(SUPPORT_EEPROM_CMD) $(LED_START_ON_CMD) $(APPSPM_CMD)
 COMMON_OPTIONS += $(VERSION_CMD)
+COMMON_OPTIONS += $(RS485_CMD)
 
 #UART is handled separately and only passed for devices with more than one.
 HELPTEXT += "Option UART=n                - use UARTn for communications\n"
diff --git a/optiboot/bootloaders/optiboot/optiboot.c b/optiboot/bootloaders/optiboot/optiboot.c
index 9a2c80b..4b9f550 100644
--- a/optiboot/bootloaders/optiboot/optiboot.c
+++ b/optiboot/bootloaders/optiboot/optiboot.c
@@ -690,6 +690,11 @@ int main(void) {
 #endif
 #endif
 
+#ifdef RS485
+  RS485_DDR |= _BV(RS485);
+  RS485_PORT &= ~_BV(RS485);
+#endif
+
   /* Forever loop: exits by causing WDT reset */
   for (;;) {
     /* get character from UART */
@@ -912,15 +917,39 @@ int main(void) {
 
 void putch(char ch) {
 #ifndef SOFT_UART
-  #ifndef LIN_UART
-    while (!(UART_SRA & _BV(UDRE0))) {  /* Spin */ }
-  #else
-    while (!(LINSIR & _BV(LTXOK)))   {  /* Spin */ }
-  #endif
-
-  UART_UDR = ch;
-
-#else
+  #ifdef RS485
+    uint8_t x;
+    #ifdef LIN_UART
+      #error LIN_UART not supported in RS485 mode
+    #endif
+    do {
+      x = UART_SRA;
+    } while (!(x & _BV(UDRE0)));
+    // clear transmitted flag
+    x |= _BV(TXC0);
+    UART_SRA = x;
+    // put transceiver to output mode
+    RS485_PORT |= _BV(RS485);
+    // put char
+    UART_UDR = ch;
+    // wait for char transmitted
+    while (!(UART_SRA & _BV(TXC0)));
+    // put transceiver to input mode
+    RS485_PORT &= ~_BV(RS485);
+  #else // ifndef RS485
+    #ifndef LIN_UART
+      while (!(UART_SRA & _BV(UDRE0))) {  /* Spin */ }
+    #else
+      while (!(LINSIR & _BV(LTXOK)))   {  /* Spin */ }
+    #endif
+
+    UART_UDR = ch;
+  #endif // ifdef RS485
+#else // ifdef SOFT_UART
+    #ifdef RS485
+      // put tranceiver in output mode
+      RS485_PORT |= _BV(RS485);
+    #endif
   __asm__ __volatile__ (
     "   com %[ch]\n" // ones complement, carry set
     "   sec\n"
@@ -943,6 +972,10 @@ void putch(char ch) {
     :
       "r25"
   );
+  #ifdef RS485
+    // put transciever in input mode
+    RS485_PORT &= ~_BV(RS485);
+  #endif
 #endif
 }
 
diff --git a/optiboot/bootloaders/optiboot/pin_defs.h b/optiboot/bootloaders/optiboot/pin_defs.h
index 9c163c4..8326000 100644
--- a/optiboot/bootloaders/optiboot/pin_defs.h
+++ b/optiboot/bootloaders/optiboot/pin_defs.h
@@ -1533,3 +1533,550 @@
 #pragma GCC diagnostic warning "-Wfatal-errors"
 #error Unrecognized LED name.  Should be like "B5"
 #endif
+
+#if RS485
+#if RS485 == B0
+#undef RS485
+#define RS485_DDR     DDRB
+#define RS485_PORT    PORTB
+#define RS485_PIN     PINB
+#define RS485         PINB0
+#elif RS485 == B1
+#undef RS485
+#define RS485_DDR     DDRB
+#define RS485_PORT    PORTB
+#define RS485_PIN     PINB
+#define RS485         PINB1
+#elif RS485 == B2
+#undef RS485
+#define RS485_DDR     DDRB
+#define RS485_PORT    PORTB
+#define RS485_PIN     PINB
+#define RS485         PINB2
+#elif RS485 == B3
+#undef RS485
+#define RS485_DDR     DDRB
+#define RS485_PORT    PORTB
+#define RS485_PIN     PINB
+#define RS485         PINB3
+#elif RS485 == B4
+#undef RS485
+#define RS485_DDR     DDRB
+#define RS485_PORT    PORTB
+#define RS485_PIN     PINB
+#define RS485         PINB4
+#elif RS485 == B5
+#undef RS485
+#define RS485_DDR     DDRB
+#define RS485_PORT    PORTB
+#define RS485_PIN     PINB
+#define RS485         PINB5
+#elif RS485 == B6
+#undef RS485
+#define RS485_DDR     DDRB
+#define RS485_PORT    PORTB
+#define RS485_PIN     PINB
+#define RS485         PINB6
+#elif RS485 == B7
+#undef RS485
+#define RS485_DDR     DDRB
+#define RS485_PORT    PORTB
+#define RS485_PIN     PINB
+#define RS485         PINB7
+
+#elif RS485 == C0
+#undef RS485
+#define RS485_DDR     DDRC
+#define RS485_PORT    PORTC
+#define RS485_PIN     PINC
+#define RS485         PINC0
+#elif RS485 == C1
+#undef RS485
+#define RS485_DDR     DDRC
+#define RS485_PORT    PORTC
+#define RS485_PIN     PINC
+#define RS485         PINC1
+#elif RS485 == C2
+#undef RS485
+#define RS485_DDR     DDRC
+#define RS485_PORT    PORTC
+#define RS485_PIN     PINC
+#define RS485         PINC2
+#elif RS485 == C3
+#undef RS485
+#define RS485_DDR     DDRC
+#define RS485_PORT    PORTC
+#define RS485_PIN     PINC
+#define RS485         PINC3
+#elif RS485 == C4
+#undef RS485
+#define RS485_DDR     DDRC
+#define RS485_PORT    PORTC
+#define RS485_PIN     PINC
+#define RS485         PINC4
+#elif RS485 == C5
+#undef RS485
+#define RS485_DDR     DDRC
+#define RS485_PORT    PORTC
+#define RS485_PIN     PINC
+#define RS485         PINC5
+#elif RS485 == C6
+#undef RS485
+#define RS485_DDR     DDRC
+#define RS485_PORT    PORTC
+#define RS485_PIN     PINC
+#define RS485         PINC6
+#elif RS485 == C7
+#undef RS485
+#define RS485_DDR     DDRC
+#define RS485_PORT    PORTC
+#define RS485_PIN     PINC
+#define RS485         PINC7
+
+#elif RS485 == D0
+#undef RS485
+#define RS485_DDR     DDRD
+#define RS485_PORT    PORTD
+#define RS485_PIN     PIND
+#define RS485         PIND0
+#elif RS485 == D1
+#undef RS485
+#define RS485_DDR     DDRD
+#define RS485_PORT    PORTD
+#define RS485_PIN     PIND
+#define RS485         PIND1
+#elif RS485 == D2
+#undef RS485
+#define RS485_DDR     DDRD
+#define RS485_PORT    PORTD
+#define RS485_PIN     PIND
+#define RS485         PIND2
+#elif RS485 == D3
+#undef RS485
+#define RS485_DDR     DDRD
+#define RS485_PORT    PORTD
+#define RS485_PIN     PIND
+#define RS485         PIND3
+#elif RS485 == D4
+#undef RS485
+#define RS485_DDR     DDRD
+#define RS485_PORT    PORTD
+#define RS485_PIN     PIND
+#define RS485         PIND4
+#elif RS485 == D5
+#undef RS485
+#define RS485_DDR     DDRD
+#define RS485_PORT    PORTD
+#define RS485_PIN     PIND
+#define RS485         PIND5
+#elif RS485 == D6
+#undef RS485
+#define RS485_DDR     DDRD
+#define RS485_PORT    PORTD
+#define RS485_PIN     PIND
+#define RS485         PIND6
+#elif RS485 == D7
+#undef RS485
+#define RS485_DDR     DDRD
+#define RS485_PORT    PORTD
+#define RS485_PIN     PIND
+#define RS485         PIND7
+
+#elif RS485 == E0
+#undef RS485
+#define RS485_DDR     DDRE
+#define RS485_PORT    PORTE
+#define RS485_PIN     PINE
+#define RS485         PINE0
+#elif RS485 == E1
+#undef RS485
+#define RS485_DDR     DDRE
+#define RS485_PORT    PORTE
+#define RS485_PIN     PINE
+#define RS485         PINE1
+#elif RS485 == E2
+#undef RS485
+#define RS485_DDR     DDRE
+#define RS485_PORT    PORTE
+#define RS485_PIN     PINE
+#define RS485         PINE2
+#elif RS485 == E3
+#undef RS485
+#define RS485_DDR     DDRE
+#define RS485_PORT    PORTE
+#define RS485_PIN     PINE
+#define RS485         PINE3
+#elif RS485 == E4
+#undef RS485
+#define RS485_DDR     DDRE
+#define RS485_PORT    PORTE
+#define RS485_PIN     PINE
+#define RS485         PINE4
+#elif RS485 == E5
+#undef RS485
+#define RS485_DDR     DDRE
+#define RS485_PORT    PORTE
+#define RS485_PIN     PINE
+#define RS485         PINE5
+#elif RS485 == E6
+#undef RS485
+#define RS485_DDR     DDRE
+#define RS485_PORT    PORTE
+#define RS485_PIN     PINE
+#define RS485         PINE6
+#elif RS485 == E7
+#undef RS485
+#define RS485_DDR     DDRE
+#define RS485_PORT    PORTE
+#define RS485_PIN     PINE
+#define RS485         PINE7
+
+#elif RS485 == F0
+#undef RS485
+#define RS485_DDR     DDRF
+#define RS485_PORT    PORTF
+#define RS485_PIN     PINF
+#define RS485         PINF0
+#elif RS485 == F1
+#undef RS485
+#define RS485_DDR     DDRF
+#define RS485_PORT    PORTF
+#define RS485_PIN     PINF
+#define RS485         PINF1
+#elif RS485 == F2
+#undef RS485
+#define RS485_DDR     DDRF
+#define RS485_PORT    PORTF
+#define RS485_PIN     PINF
+#define RS485         PINF2
+#elif RS485 == F3
+#undef RS485
+#define RS485_DDR     DDRF
+#define RS485_PORT    PORTF
+#define RS485_PIN     PINF
+#define RS485         PINF3
+#elif RS485 == F4
+#undef RS485
+#define RS485_DDR     DDRF
+#define RS485_PORT    PORTF
+#define RS485_PIN     PINF
+#define RS485         PINF4
+#elif RS485 == F5
+#undef RS485
+#define RS485_DDR     DDRF
+#define RS485_PORT    PORTF
+#define RS485_PIN     PINF
+#define RS485         PINF5
+#elif RS485 == F6
+#undef RS485
+#define RS485_DDR     DDRF
+#define RS485_PORT    PORTF
+#define RS485_PIN     PINF
+#define RS485         PINF6
+#elif RS485 == F7
+#undef RS485
+#define RS485_DDR     DDRF
+#define RS485_PORT    PORTF
+#define RS485_PIN     PINF
+#define RS485         PINF7
+
+#elif RS485 == G0
+#undef RS485
+#define RS485_DDR     DDRG
+#define RS485_PORT    PORTG
+#define RS485_PIN     PING
+#define RS485         PING0
+#elif RS485 == G1
+#undef RS485
+#define RS485_DDR     DDRG
+#define RS485_PORT    PORTG
+#define RS485_PIN     PING
+#define RS485         PING1
+#elif RS485 == G2
+#undef RS485
+#define RS485_DDR     DDRG
+#define RS485_PORT    PORTG
+#define RS485_PIN     PING
+#define RS485         PING2
+#elif RS485 == G3
+#undef RS485
+#define RS485_DDR     DDRG
+#define RS485_PORT    PORTG
+#define RS485_PIN     PING
+#define RS485         PING3
+#elif RS485 == G4
+#undef RS485
+#define RS485_DDR     DDRG
+#define RS485_PORT    PORTG
+#define RS485_PIN     PING
+#define RS485         PING4
+#elif RS485 == G5
+#undef RS485
+#define RS485_DDR     DDRG
+#define RS485_PORT    PORTG
+#define RS485_PIN     PING
+#define RS485         PING5
+#elif RS485 == G6
+#undef RS485
+#define RS485_DDR     DDRG
+#define RS485_PORT    PORTG
+#define RS485_PIN     PING
+#define RS485         PING6
+#elif RS485 == G7
+#undef RS485
+#define RS485_DDR     DDRG
+#define RS485_PORT    PORTG
+#define RS485_PIN     PING
+#define RS485         PING7
+
+#elif RS485 == H0
+#undef RS485
+#define RS485_DDR     DDRH
+#define RS485_PORT    PORTH
+#define RS485_PIN     PINH
+#define RS485         PINH0
+#elif RS485 == H1
+#undef RS485
+#define RS485_DDR     DDRH
+#define RS485_PORT    PORTH
+#define RS485_PIN     PINH
+#define RS485         PINH1
+#elif RS485 == H2
+#undef RS485
+#define RS485_DDR     DDRH
+#define RS485_PORT    PORTH
+#define RS485_PIN     PINH
+#define RS485         PINH2
+#elif RS485 == H3
+#undef RS485
+#define RS485_DDR     DDRH
+#define RS485_PORT    PORTH
+#define RS485_PIN     PINH
+#define RS485         PINH3
+#elif RS485 == H4
+#undef RS485
+#define RS485_DDR     DDRH
+#define RS485_PORT    PORTH
+#define RS485_PIN     PINH
+#define RS485         PINH4
+#elif RS485 == H5
+#undef RS485
+#define RS485_DDR     DDRH
+#define RS485_PORT    PORTH
+#define RS485_PIN     PINH
+#define RS485         PINH5
+#elif RS485 == H6
+#undef RS485
+#define RS485_DDR     DDRH
+#define RS485_PORT    PORTH
+#define RS485_PIN     PINH
+#define RS485         PINH6
+#elif RS485 == H7
+#undef RS485
+#define RS485_DDR     DDRH
+#define RS485_PORT    PORTH
+#define RS485_PIN     PINH
+#define RS485         PINH7
+
+#elif RS485 == J0
+#undef RS485
+#define RS485_DDR     DDRJ
+#define RS485_PORT    PORTJ
+#define RS485_PIN     PINJ
+#define RS485         PINJ0
+#elif RS485 == J1
+#undef RS485
+#define RS485_DDR     DDRJ
+#define RS485_PORT    PORTJ
+#define RS485_PIN     PINJ
+#define RS485         PINJ1
+#elif RS485 == J2
+#undef RS485
+#define RS485_DDR     DDRJ
+#define RS485_PORT    PORTJ
+#define RS485_PIN     PINJ
+#define RS485         PINJ2
+#elif RS485 == J3
+#undef RS485
+#define RS485_DDR     DDRJ
+#define RS485_PORT    PORTJ
+#define RS485_PIN     PINJ
+#define RS485         PINJ3
+#elif RS485 == J4
+#undef RS485
+#define RS485_DDR     DDRJ
+#define RS485_PORT    PORTJ
+#define RS485_PIN     PINJ
+#define RS485         PINJ4
+#elif RS485 == J5
+#undef RS485
+#define RS485_DDR     DDRJ
+#define RS485_PORT    PORTJ
+#define RS485_PIN     PINJ
+#define RS485         PINJ5
+#elif RS485 == J6
+#undef RS485
+#define RS485_DDR     DDRJ
+#define RS485_PORT    PORTJ
+#define RS485_PIN     PINJ
+#define RS485         PINJ6
+#elif RS485 == J7
+#undef RS485
+#define RS485_DDR     DDRJ
+#define RS485_PORT    PORTJ
+#define RS485_PIN     PINJ
+#define RS485         PINJ7
+
+#elif RS485 == K0
+#undef RS485
+#define RS485_DDR     DDRK
+#define RS485_PORT    PORTK
+#define RS485_PIN     PINK
+#define RS485         PINK0
+#elif RS485 == K1
+#undef RS485
+#define RS485_DDR     DDRK
+#define RS485_PORT    PORTK
+#define RS485_PIN     PINK
+#define RS485         PINK1
+#elif RS485 == K2
+#undef RS485
+#define RS485_DDR     DDRK
+#define RS485_PORT    PORTK
+#define RS485_PIN     PINK
+#define RS485         PINK2
+#elif RS485 == K3
+#undef RS485
+#define RS485_DDR     DDRK
+#define RS485_PORT    PORTK
+#define RS485_PIN     PINK
+#define RS485         PINK3
+#elif RS485 == K4
+#undef RS485
+#define RS485_DDR     DDRK
+#define RS485_PORT    PORTK
+#define RS485_PIN     PINK
+#define RS485         PINK4
+#elif RS485 == K5
+#undef RS485
+#define RS485_DDR     DDRK
+#define RS485_PORT    PORTK
+#define RS485_PIN     PINK
+#define RS485         PINK5
+#elif RS485 == K6
+#undef RS485
+#define RS485_DDR     DDRK
+#define RS485_PORT    PORTK
+#define RS485_PIN     PINK
+#define RS485         PINK6
+#elif RS485 == K7
+#undef RS485
+#define RS485_DDR     DDRK
+#define RS485_PORT    PORTK
+#define RS485_PIN     PINK
+#define RS485         PINK7
+
+#elif RS485 == L0
+#undef RS485
+#define RS485_DDR     DDRL
+#define RS485_PORT    PORTL
+#define RS485_PIN     PINL
+#define RS485         PINL0
+#elif RS485 == L1
+#undef RS485
+#define RS485_DDR     DDRL
+#define RS485_PORT    PORTL
+#define RS485_PIN     PINL
+#define RS485         PINL1
+#elif RS485 == L2
+#undef RS485
+#define RS485_DDR     DDRL
+#define RS485_PORT    PORTL
+#define RS485_PIN     PINL
+#define RS485         PINL2
+#elif RS485 == L3
+#undef RS485
+#define RS485_DDR     DDRL
+#define RS485_PORT    PORTL
+#define RS485_PIN     PINL
+#define RS485         PINL3
+#elif RS485 == L4
+#undef RS485
+#define RS485_DDR     DDRL
+#define RS485_PORT    PORTL
+#define RS485_PIN     PINL
+#define RS485         PINL4
+#elif RS485 == L5
+#undef RS485
+#define RS485_DDR     DDRL
+#define RS485_PORT    PORTL
+#define RS485_PIN     PINL
+#define RS485         PINL5
+#elif RS485 == L6
+#undef RS485
+#define RS485_DDR     DDRL
+#define RS485_PORT    PORTL
+#define RS485_PIN     PINL
+#define RS485         PINL6
+#elif RS485 == L7
+#undef RS485
+#define RS485_DDR     DDRL
+#define RS485_PORT    PORTL
+#define RS485_PIN     PINL
+#define RS485         PINL7
+
+#elif RS485 == A0
+#undef RS485
+#define RS485_DDR     DDRA
+#define RS485_PORT    PORTA
+#define RS485_PIN     PINA
+#define RS485         PINA0
+#elif RS485 == A1
+#undef RS485
+#define RS485_DDR     DDRA
+#define RS485_PORT    PORTA
+#define RS485_PIN     PINA
+#define RS485         PINA1
+#elif RS485 == A2
+#undef RS485
+#define RS485_DDR     DDRA
+#define RS485_PORT    PORTA
+#define RS485_PIN     PINA
+#define RS485         PINA2
+#elif RS485 == A3
+#undef RS485
+#define RS485_DDR     DDRA
+#define RS485_PORT    PORTA
+#define RS485_PIN     PINA
+#define RS485         PINA3
+#elif RS485 == A4
+#undef RS485
+#define RS485_DDR     DDRA
+#define RS485_PORT    PORTA
+#define RS485_PIN     PINA
+#define RS485         PINA4
+#elif RS485 == A5
+#undef RS485
+#define RS485_DDR     DDRA
+#define RS485_PORT    PORTA
+#define RS485_PIN     PINA
+#define RS485         PINA5
+#elif RS485 == A6
+#undef RS485
+#define RS485_DDR     DDRA
+#define RS485_PORT    PORTA
+#define RS485_PIN     PINA
+#define RS485         PINA6
+#elif RS485 == A7
+#undef RS485
+#define RS485_DDR     DDRA
+#define RS485_PORT    PORTA
+#define RS485_PIN     PINA
+#define RS485         PINA7
+
+#else
+#error -------------------------------------------
+#error Unrecognized RS485 name.  Should be like "B5"
+#error -------------------------------------------
+#endif
+#endif
-- 
2.7.4

